File: leetcode/1-100/20-valid-parentheses/bench/valid_parentheses.py
"""Benchmark workload — Valid Parentheses (LeetCode #20).

Python mirror of bench/valid_parentheses.kara. Same byte-buffer build +
stack-of-expected-closers validate + count-valid sink. Runs a scaled-down
K = 100_000 (the compiled mirrors run K = 500_000) — timed separately, not
in the cross-mirror sink check. Sink at K = 100_000: corrupt fires on
k % 7 == 0 (⌈100000/7⌉ = 14286 iters), so valid = 100000 − 14286 = 85714.
See ../README.md § Benchmarks.
"""

from __future__ import annotations

_CLOSER = {ord("("): ord(")"), ord("["): ord("]"), ord("{"): ord("}")}


def is_valid_bytes(data: bytes) -> bool:
    stack: list[int] = []
    for b in data:
        c = _CLOSER.get(b)
        if c is not None:
            stack.append(c)
        else:
            if not stack or stack.pop() != b:
                return False
    return not stack


def build_brackets(depth: int, kind: int, corrupt: bool) -> bytes:
    op, cl, wrong = (ord("("), ord(")"), ord("]"))
    if kind == 1:
        op, cl, wrong = (ord("["), ord("]"), ord(")"))
    elif kind == 2:
        op, cl, wrong = (ord("{"), ord("}"), ord(")"))
    buf = bytearray()
    buf.extend(op for _ in range(depth))
    buf.extend(cl for _ in range(depth - 1))
    buf.append(wrong if corrupt else cl)
    return bytes(buf)


def main() -> None:
    depth = 1000
    k_iters = 100_000

    count = 0
    for k in range(k_iters):
        kind = k % 3
        corrupt = (k % 7) == 0
        buf = build_brackets(depth, kind, corrupt)
        if is_valid_bytes(buf):
            count += 1
    print(count)


if __name__ == "__main__":
    main()

File: leetcode/1-100/20-valid-parentheses/bench/valid_parentheses.rs
// Benchmark workload — Valid Parentheses (LeetCode #20).
// Rust mirror of bench/valid_parentheses.kara. Same byte-buffer build +
// stack-of-expected-closers validate + count-valid sink, single-threaded.
// `Vec<u8>` grown from empty (no pre-reserve) mirrors Kāra's `Vec.new()`
// allocator traffic; the stack is a fresh `Vec<u8>` per call. Compiled
// with `rustc -O`. The seq lane's honest single-thread comparison.
// See ../README.md § Benchmarks.

use std::hint::black_box;

fn is_valid_bytes(bytes: &[u8]) -> bool {
    let mut stack: Vec<u8> = Vec::new();
    for &b in bytes {
        if b == b'(' || b == b'[' || b == b'{' {
            stack.push(match b {
                b'(' => b')',
                b'[' => b']',
                _ => b'}',
            });
        } else {
            match stack.pop() {
                Some(top) => {
                    if top != b {
                        return false;
                    }
                }
                None => return false,
            }
        }
    }
    stack.is_empty()
}

// `2*depth`-byte nested bracket buffer of the given kind; when `corrupt`,
// the final closer is flipped to a different closer type (invalid).
fn build_brackets(depth: i64, kind: i64, corrupt: bool) -> Vec<u8> {
    let (op, cl, wrong) = match kind {
        1 => (b'[', b']', b')'),
        2 => (b'{', b'}', b')'),
        _ => (b'(', b')', b']'),
    };
    let mut buf: Vec<u8> = Vec::new();
    for _ in 0..depth {
        buf.push(op);
    }
    for _ in 0..depth - 1 {
        buf.push(cl);
    }
    buf.push(if corrupt { wrong } else { cl });
    buf
}

fn main() {
    let depth: i64 = 1000;
    let k_iters: i64 = 500_000;

    let mut count: i64 = 0;
    for k in 0..k_iters {
        let kind = k % 3;
        let corrupt = (k % 7) == 0;
        let buf = build_brackets(depth, kind, corrupt);
        if is_valid_bytes(black_box(&buf)) {
            count += 1;
        }
    }
    println!("{}", count);
}
